src/lib/credits.ts

```diff
@@ -0,0 +1,625 @@
+  prepareCreditsLegacyMarketplace(
+    nft: NFT,
+    order: Order,
+    credits: Credit[]
+  ): {
+    contract: ContractData
+    creditsData: CreditsData[]
+    creditsSignatures: string[]
+    externalCall: ExternalCallParams
+    maxUncreditedValue: string
+    maxCreditedValue: string
+  } {
+    // Prepare common credits data
+    const {
+      contract,
+      creditsData,
+      creditsSignatures
+    } = this.prepareCreditsData(credits, nft.chainId)
+
+    console.log('creditsData', creditsData) // TODO: remove
+    console.log('creditsSignatures', creditsSignatures) // TODO: remove
+
+    // Get the Marketplace contract address
+    // const marketplaceContract = getContract(ContractName.Marketplace, nft.chainId)
+    const marketplaceContract = getContract(
+      nft.network === Network.ETHEREUM
+        ? ContractName.Marketplace
+        : ContractName.MarketplaceV2,
+      nft.chainId
+    )
+    const marketplaceAddress = marketplaceContract.address
+
+    // Create a contract interface for the Marketplace to get the function selector
+    const marketplaceInterface = new ethers.utils.Interface(
+      marketplaceContract.abi
+    )
+
+    // The selector for the executeOrder function in the Marketplace contract
+    const executeOrderSelector = marketplaceInterface.getSighash('executeOrder')
+
+    // Encode the executeOrder function parameters
+    const executeOrderData = ethers.utils.defaultAbiCoder.encode(
+      ['address', 'uint256', 'uint256'],
+      [nft.contractAddress, nft.tokenId, order.price]
+    )
+
+    // Prepare the external call
+    const externalCall = this.prepareExternalCall({
+      target: marketplaceAddress,
+      selector: executeOrderSelector,
+      data: executeOrderData
+      // chainId: nft.chainId
+    })
+
+    console.log('externalCall', externalCall)
+    const creditsValue = credits.reduce(
+      (acc, credit) => acc + parseInt(credit.amount),
+      0
+    )
+    const whatUserHasToPay = BigInt(order.price) - BigInt(creditsValue)
+    console.log('whatUserHasToPay', whatUserHasToPay)
+    const maxUncreditedValue =
+      whatUserHasToPay < BigInt(0) ? '0' : whatUserHasToPay.toString()
+    console.log('maxUncreditedValue', maxUncreditedValue)
+
+    return {
+      contract,
+      creditsData,
+      creditsSignatures,
+      externalCall,
+      maxUncreditedValue,
+      maxCreditedValue: order.price
+    }
+  }
+
+  /**
+   * Use credits for buying NFTs from the Legacy Marketplace
+   * @param nft - The NFT to buy
+   * @param order - The order to execute
+   * @param credits - The user's credits
+   * @returns The transaction hash
+   */
+  async useCreditsLegacyMarketplace(
+    nft: NFT,
+    order: Order,
+    credits: Credit[]
+  ): Promise<string> {
+    // Prepare common credits data
+    const {
+      contract,
+      creditsData,
+      creditsSignatures
+    } = this.prepareCreditsData(credits, nft.chainId)
+
+    console.log('creditsData', creditsData) // TODO: remove
```

Comment:
The `prepareCreditsLegacyMarketplace` and `useCreditsLegacyMarketplace` have code in common, would you mind using `prepareCreditsLegacyMarketplace` into `useCreditsLegacyMarketplace` or abstract the code so both function can benefit from it?